src/app/services/ninja-service.ts

```diff
@@ -87,21 +87,34 @@ export class NinjaService {
     pathwayPrediction: PathwayPredictionResult,
     reactions: PathwayPredictionReactions,
     metabolites: PathwayPredictionMetabolites): Observable<AddedReaction[]> {
-    const heterologousReactions = pathwayPrediction.heterologous_reactions.map((reactionId) => {
-      const metaboliteIds = Object.keys(reactions[reactionId].metabolites);
-      return this.mapMnxMetabolitesToBigg(metaboliteIds).pipe(map((ids) => {
+    const addedReactionIds = [...pathwayPrediction.heterologous_reactions, ...pathwayPrediction.synthetic_reactions];
+    const metabolitesInModel = new Set(pathwayPrediction.model.model_serialized.metabolites.map((metabolite) => metabolite.id));
+
+    const addedReactions = addedReactionIds.map((reactionId) => {
+      const mnxMetabolitesInReaction = reactions[reactionId].metabolites;
+      const mnxMetaboliteIds = Object.keys(mnxMetabolitesInReaction);
+      return this.mapMnxMetabolitesToBigg(mnxMetaboliteIds).pipe(map((ids) => {
         const metabolites_to_add = [];
         const biggMetabolites = {};
-        const mnxMetabolites = reactions[reactionId].metabolites;
-          for (const mnxId in ids) {
-            if (mnxId) {
-              const biggId = ids[mnxId][0] + '_c';
-              const metabolite = metabolites[mnxId];
-              metabolite.id = biggId;
-              metabolites_to_add.push(metabolite);
-              biggMetabolites[biggId] = mnxMetabolites[mnxId];
+        for (const mnxId of Object.keys(ids)) {
+          const mappedBiggIds = ids[mnxId];
+          let isMetaboliteInModel = false;
+          let biggId;
+          // if id-mapper returns more than 1 mapped bigg id, take the one that is in the model
+          mappedBiggIds.forEach((id) => {
+            if (metabolitesInModel.has(id + '_c')) {
+              isMetaboliteInModel = true;
+              biggId = id + '_c';
             }
+          });
+          if (!isMetaboliteInModel) {
+            biggId = mappedBiggIds[0];
```

Comment:
Should we deal with the case if there is no match result returned from the id-mapper?

Reply:
If I remembered properly, `ids[mnxId]` always have elements, so, we should not needed, but, just in case, you can add the operator  || here `biggId = mappedBiggIds[0] || something`;

Reply:
If id-mapper didn't match the mnx id to bigg, it doesn't return mnxId key in ids object. So to consider these cases now we are still adding such metabolite but we keep its mnx id.